kernel/cClosure.ml

```diff
@@ -485,55 +485,53 @@ let subst_context e ctx =
   snd @@ subst_context ctx
 
 (* The inverse of mk_clos: move back to constr *)
-(* XXX should there be universes in lfts???? *)
-let rec to_constr (lfts, usubst as ulfts) v =
-  let subst_us c = subst_instance_constr usubst c in
+let rec to_constr lfts v =
   match v.term with
     | FRel i -> mkRel (reloc_rel i lfts)
     | FFlex (RelKey p) -> mkRel (reloc_rel p lfts)
     | FFlex (VarKey x) -> mkVar x
-    | FAtom c -> subst_us (exliftn lfts c)
-    | FFlex (ConstKey op) -> subst_us (mkConstU op)
-    | FInd op -> subst_us (mkIndU op)
-    | FConstruct op -> subst_us (mkConstructU op)
+    | FAtom c -> exliftn lfts c
+    | FFlex (ConstKey op) -> mkConstU op
+    | FInd op -> mkIndU op
+    | FConstruct op -> mkConstructU op
     | FCaseT (ci, u, pms, p, c, ve, env) ->
-      to_constr_case ulfts ci u pms p NoInvert c ve env
+      to_constr_case lfts ci u pms p NoInvert c ve env
     | FCaseInvert (ci, u, pms, p, indices, c, ve, env) ->
-      let iv = CaseInvert {indices=Array.Fun1.map to_constr ulfts indices} in
-      to_constr_case ulfts ci u pms p iv c ve env
+      let iv = CaseInvert {indices=Array.Fun1.map to_constr lfts indices} in
+      to_constr_case lfts ci u pms p iv c ve env
     | FFix ((op,(lna,tys,bds)) as fx, e) ->
       if is_subs_id (fst e) && is_lift_id lfts then
-        subst_instance_constr (usubst_instance ulfts (snd e)) (mkFix fx)
+        subst_instance_constr (snd e) (mkFix fx)
       else
         let n = Array.length bds in
-        let subs_ty = comp_subs ulfts e in
-        let subs_bd = comp_subs (on_fst (el_liftn n) ulfts) (on_fst (subs_liftn n) e) in
+        let subs_ty = comp_subs lfts e in
+        let subs_bd = comp_subs (el_liftn n lfts) (on_fst (subs_liftn n) e) in
```

Comment:
Is this equivalent to
```suggestion
        let subs_bd = el_liftn n subs_ty in
```
?
in which case it should be slightly more efficient to write it that way to share the lazy `to_constr` generated by comp_subs
(although I doubt any real code needs efficient CClosure.to_constr on raw fixpoints)

Reply:
I hope that's the case, otherwise there is something wrong with to_constr. Even without applying this optimization, a small comment stating the explicit typing rule and equational properties of to_constr in the style of the Esubst mli would be welcome.